includes/class-wp-job-manager-post-types.php

```diff
@@ -659,6 +660,83 @@ public function job_feed() {
 		remove_filter( 'posts_search', 'get_job_listings_keyword_search' );
 	}
 
+	/**
+	 * Get filled jobs.
+	 *
+	 * @return array
+	 */
+	public function get_filled_job_listings(): array {
+		if ( ! get_option( 'job_manager_hide_filled_positions' ) ) {
+			return [];
+		}
+
+		$filled_jobs_transient = get_transient( 'hide_filled_jobs_transient' );
+		if ( false === $filled_jobs_transient ) {
+			$filled_jobs_transient = get_posts(
+				[
+					'post_type'  => 'job_listing',
+					'fields'     => 'ids',
+					'meta_query' => [
+						[
+							'key'     => '_filled',
+							'value'   => '1',
+							'compare' => '=',
+						],
+					],
+				]
+			);
+			set_transient( 'hide_filled_jobs_transient', $filled_jobs_transient, DAY_IN_SECONDS );
+		}
+		return $filled_jobs_transient;
+	}
+
+	/**
+	 * Get expired jobs.
+	 *
+	 * @return array
+	 */
+	public function get_expired_job_listings(): array {
+		if ( ! get_option( 'job_manager_hide_expired' ) ) {
+			return [];
+		}
+
+		$expired_jobs_transient = get_transient( 'hide_expired_jobs_transient' );
+		if ( false === $expired_jobs_transient ) {
+			$expired_jobs_transient = get_posts(
+				[
+					'post_type'   => 'job_listing',
+					'post_status' => 'expired',
+					'fields'      => 'ids',
+				]
+			);
+			set_transient( 'hide_expired_jobs_transient', $expired_jobs_transient, DAY_IN_SECONDS );
+		}
+		return $expired_jobs_transient;
+	}
+
+	/**
+	 * Maybe exclude expired and/or filled job listings from search.
+	 *
+	 * @param $query WP_Query $query
+	 *
+	 * @return void
+	 */
+	public function maybe_hide_filled_expired_job_listings_from_search( WP_Query $query ): void {
+		$hide_filled_positions = get_option( 'job_manager_hide_filled_positions' );
+		$hide_expired          = get_option( 'job_manager_hide_expired' );
+
+		if ( ! $hide_filled_positions && ! $hide_expired ) {
+			return;
+		}
+
+		if ( ! is_admin() && $query->is_search() ) {
```

Comment:
We should add a post type check `'job_listing' === $query->get( 'post_type' )`, and also best to add `$query->is_main_query()`.

It'd be nice if this also worked for the archive pages (docs: https://wpjobmanager.com/document/enabling-full-template-support/, recent report: p1687165209914009-slack-C6GGX896G). The check for that would be `$query->is_archive()`

For the expired listings, do you think it's worth doing the transient and dedicated query? Seems like it would be simpler to filter the post status on the main query with something like `$query->set( 'post_status', 'publish' );` 

Also I'm not sure if we have some other filter adding back the `expired` status for this query, as it's [registered with `'exclude_from_search' => true`](https://github.com/Automattic/WP-Job-Manager/blob/38019efba98ce55b2c881f30003a3b8059f9e1e1/includes/class-wp-job-manager-post-types.php#L395). In WP_Query that's respected when the queried post status is `'any'` — might be that one is not set by default for search/archive queries?

Reply:
@yscik 

> We should add a post type check 'job_listing' === $query->get( 'post_type' ), and also best to add $query->is_main_query().

> It'd be nice if this also worked for the archive pages (docs: https://wpjobmanager.com/document/enabling-full-template-support/, recent report: p1687165209914009-slack-C6GGX896G). The check for that would be $query->is_archive()

Ah right, to avoid secondary (or other) queries from being impacted! 

> For the expired listings, do you think it's worth doing the transient and dedicated query? Seems like it would be simpler to filter the post status on the main query with something like $query->set( 'post_status', 'publish' );

That... would be far simpler. Too much tunnel vision.

> Also I'm not sure if we have some other filter adding back the expired status for this query, as it's [registered with 'exclude_from_search' => true](https://github.com/Automattic/WP-Job-Manager/blob/38019efba98ce55b2c881f30003a3b8059f9e1e1/includes/class-wp-job-manager-post-types.php#L395). In WP_Query that's respected when the queried post status is 'any' — might be that one is not set by default for search/archive queries?

You would think, but there's a bug where this isn't the behaviour that happens https://core.trac.wordpress.org/ticket/44737. As a result, what should be a simple `exclude_from_search` isn't. It's possible to get this option to work by changing the parameter on the post type registration:

https://github.com/Automattic/WP-Job-Manager/blob/a09f7fad82dd3b7080a4c848fc08d91afbaf5755/includes/class-wp-job-manager-post-types.php#L369

But this isn't great, and @aaronfc and I decided to go in the direction we did originally, which I've continued with here.

One other thing I noticed is that the `get_filled_job_listings()` method was limited to something like 5 posts without the `posts_per_page` parameter, so I've set that hesitantly to `-1` and kept in the transient code for it (but removed for expired posts). I'm not aware of a way to avoid this as the `_filled` meta is, well, meta. And, if we're excluding all posts, we need all posts...

I'm not sure if maybe it would be worth running the `meta_query` directly on the `$query->set`, but my feeling is that this then borks any attempt at caching:

```
$query->set( 'meta_query', array(
        array(
            'key'     => '_filled',
            'value'   => '1',
            'compare' => '!='
        ),
    )
);
```

Changes committed here:

- https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/d1d04d450117cb0c207f0aa4476dc63b8b2206f0
- https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/3fedb2bc7c13d6254316e4a6a0dd4e69be6ccdb1